File: 234-palindrome-linked-list/234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode node=head;
        int sz=0;
        while(node!=null){
            node=node.next;
            sz++;
        }
        if(sz==1)return true;
        ListNode curr=head;
        ListNode prev=null;
        for(int i=0;i<sz/2;i++){
            ListNode next=curr.next;
            curr.next=prev;
            prev=curr;
            curr=next;
        }
        if(sz%2!=0){
            curr=curr.next;
        }
        while(prev!=null || curr!=null){
            if(prev.val!=curr.val)return false;
            prev=prev.next;
            curr=curr.next;
        }
        if(curr==null && prev==null)return true;
        return false;
    }
}

File: 234-palindrome-linked-list/234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return True
        slow = head
        fast = head
        while fast.next is not None and fast.next.next is not None:
            slow = slow.next
            fast = fast.next.next
        mid = slow.next
        slow.next = None
        prev = None
        while mid is not None:
            temp = mid.next
            mid.next = prev
            prev = mid
            mid = temp
        while prev is not None and head is not None:
            if prev.val != head.val:
                return False
            prev = prev.next
            head = head.next
        return True